core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Resources;
+import com.google.common.reflect.ClassPath;
+import com.google.common.reflect.ClassPath.ClassInfo;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Properties;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+import static java.lang.ClassLoader.getSystemClassLoader;
+import static java.lang.Integer.parseInt;
+import static java.lang.String.format;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestSingleVersionBackwardCompatibility
+{
+    @Test
+    public void test()
+            throws Exception
+    {
+        assertThat(getCurrentVersionEntities()).containsAll(getPreviousVersionEntities());
+    }
+
+    private static List<String> getCurrentVersionEntities()
+            throws IOException
+    {
+        return getSpiEntities(getSystemClassLoader(), true);
+    }
+
+    private static List<String> getPreviousVersionEntities()
+            throws Exception
+    {
+        String previousVersion = Integer.toString(parseInt(getCurrentVersion().replace("-SNAPSHOT", "")) - 1);
+        URL slice = download(format("https://repo1.maven.org/maven2/io/airlift/slice/%1$s/slice-%1$s.jar", getSliceVersion()));
```

Comment:
Are we assuming that Slice version doesn't change?

Reply:
It can change, however I assume that loading slice classes for SPI for previous version will still work. Otherwise it sounds like yet another backward incompatible change.

Reply:
I am not sure i agree, but please capture your reasoning as a code comment